[Won Ye Ji] iP

wonyeji opened the pull request:
### Sun
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes/))

Sun frees your mind of having to remember things you need to do! It's,

- Text-based
- Easy to learn
- ~~FAST~~ _SUPER_ FAST to use

All you need to do is,
1. Download it from [here](https://github.com/wonyeji/ip/releases/tag/A-Release).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉

What's more, it is **FREE**!

Features:
- [x]  Managing tasks
- [x]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

If you Java programmer, you can use it to practice Java too! Here's the `main` method:

``` Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

wonyeji committed:
Level-1

wonyeji committed:
Level-1

wonyeji committed:
Level-2

wonyeji committed:
Level-3

wonyeji committed:
todo implemented (L4)

wonyeji committed:
Level-4

wonyeji committed:
Level-6

wonyeji committed:
Level-5

wonyeji committed:
Level-6

wonyeji committed:
A-TextUiTesting

wonyeji committed:
Clean code from previous submission

wonyeji committed:
Update exception class

wonyeji committed:
Fix newline bug

wonyeji committed:
Update DukeException

wonyeji committed:
Clean code for Week 2 IP

wonyeji committed:
Level-7

wonyeji committed:
Level-8

wonyeji committed:
Merge branch 'branch-Level-7'

wonyeji committed:
Merge branch 'branch-Level-8'

wonyeji committed:
Add Ui class

wonyeji committed:
Update Storage class

wonyeji committed:
Add Parser class

wonyeji committed:
Add Command classes

wonyeji committed:
Organise into packages

wonyeji committed:
Add Junit Test

wonyeji committed:
Add Jar file

wonyeji committed:
Add Javadocs

wonyeji committed:
Follow coding standard

wonyeji committed:
Add Find

wonyeji committed:
Handle more errors

wonyeji committed:
Update JavaDoc

wonyeji committed:
Update Coding Standard

wonyeji committed:
Update Find Command

wonyeji committed:
Merge branch 'branch-A-JavaDoc'

wonyeji committed:
Merge branch 'branch-A-CodingStandard'

wonyeji committed:
Merge branch 'branch-Level-9'

lizchow commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,23 @@
+        parser.initialiseDuke();
+        parser.runDuke();
+    }
+}
```
I think this file is missing a newline at the end of the file? Noticed it for a few other files too.

lizchow commented on src/main/java/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    /**
+     * Execute the "Done" Command.
```
```suggestion
     * Executes the "Done" Command.
```

lizchow commented on src/main/java/duke/data/task/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+public class Deadline extends Task {
+
+    protected String by;
+    private final DateTimeFormatter input = DateTimeFormatter.ofPattern("dd/MM/yy HHmm");
+    private final DateTimeFormatter output = DateTimeFormatter.ofPattern("MMM dd h:mm a");
```
Should these be renamed to `inputFormatter` and `outputFormatter` as these are not really `input` and `output`?

lizchow commented on src/main/java/duke/data/task/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+            LocalDateTime date = LocalDateTime.parse(by, input);
+            return date.format(output);
+        } catch (DateTimeParseException e) {
+            return e.getMessage();
```
Rather than returning `e.getMessage()`, I think it will be better to throw `DukeException` here as it is an error?

lizchow commented on src/main/java/duke/data/task/Task.java:
```diff
@@ -0,0 +1,54 @@
+     * @return string representation of the done status of the task.
+     */
+    public String getStatusIcon() {
+        return (isDone ? "X" : " "); // mark done task with X
```
I think the parentheses are redundant?
```suggestion
        return isDone ? "X" : " "; // mark done task with X
```

lizchow commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,98 @@
+                    ListCommand lc = new ListCommand(taskHandler, storage);
+                    lc.execute(cmd);
+                    break;
+
```
I dont think a newline is needed after a `break` statement? Same for all the other cases below.
```suggestion

```

lizchow commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,111 @@
+    public Task taskFormatter(String task) throws DukeException {
+        String type = task.substring(1, 2);
+        String doneStatus = task.substring(4, 5);
+        Task t;
```
Instead of naming it `t`, perhaps a more appropriate name will be `newTask`?
```suggestion
        Task newTask;
```

lizchow commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,135 @@
+     * @param s Type of task input.
+     * @return The error message to the user.
+     */
+    public static String emptyDescription(String s) {
```
Perhaps a better name for this parameter `s` can be `taskType`?
```suggestion
    public static String emptyDescription(String taskType) {
```

lizchow commented on src/test/java/duke/data/task/TaskTest.java:
```diff
@@ -0,0 +1,15 @@
+    @Test
+    public void markAsDone_success(){
+        Task todo = new ToDo("read book");
+        assertEquals("[T][ ] read book", todo.toString());
```
Rather than asserting `todo.toString()`, it might be more appropriate to assert the `todo.isDone()` is `false` as the `toString()` messages can change during development and it might cause tests to fail if you forget to update the tests.

lizchow commented on src/test/java/duke/data/task/TaskTest.java:
```diff
@@ -0,0 +1,15 @@
+        Task todo = new ToDo("read book");
+        assertEquals("[T][ ] read book", todo.toString());
+        todo.markAsDone();
+        assertEquals("[T][X] read book", todo.toString());
```
Same here, might be better to `assertEquals(todo.isDone(), true)`

weenleen commented on src/main/java/duke/data/TaskHandler.java:
```diff
@@ -0,0 +1,143 @@
+        list.add(event);
+        Ui.addTask();
+        System.out.println(Ui.indentation() + Ui.indentation() + event);
+    }
```
Instead of having 3 different methods to add the different types of Tasks, maybe we can make use of inheritance to have a single `addTask(Task task)` method?

weenleen commented on src/main/java/duke/data/task/Task.java:
```diff
@@ -0,0 +1,54 @@
+     *
+     * @return String representation of the task.
+     */
+    public abstract String toSave();
```
Maybe this method name can be more descriptive, or include a verb?

weenleen commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,98 @@
+                switch (commandWord) {
+                case ListCommand.COMMAND_WORD:
+                    ListCommand lc = new ListCommand(taskHandler, storage);
+                    lc.execute(cmd);
```
This line is repeated in every case, so maybe you could consider putting it outside the switch case? Just a suggestion:

```suggestion
    Command c;
    switch (commandWord) {
    // insert the cases here
    // e.g. c = new ListCommand(taskHandler, storage);
    default:
        // throw exception
    }
    c.execute(cmd);
```

weenleen commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,111 @@
+        String doneStatus = task.substring(4, 5);
+        Task t;
+        switch (type) {
+        case "D":
```
Instead of just putting the character here, perhaps you could add a final static character to the different Task classes, so that your code might be more readable / easier to edit in the future. Maybe something like `DEADLINE_SYMBOL = "D"`?

weenleen commented on src/main/java/duke/storage/Storage.java:
```diff
@@ -0,0 +1,111 @@
+        Task t;
+        switch (type) {
+        case "D":
+            String d = task.substring(7, task.length() - 1);
```
The `7` here may be considered a "magic number", so maybe you would want to add it as a constant instead? This can also inform others on what the number represents

weenleen reviewed (commented):
Good job! I think the overall code quality is pretty decent :)

wonyeji commented:
mmm

yzhilong commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,32 @@
+     *
+     * @param cmd Command string to be executed.
+     */
+    public abstract void execute(String cmd);
```
```suggestion
    public abstract void execute(String command);
```
Would it be clearer to the reader what this String represents if it was not abbreviated?

wonyeji committed:
Merge branch 'add-gradle-support'

yzhilong commented on src/main/java/duke/data/TaskHandler.java:
```diff
@@ -0,0 +1,143 @@
+package duke.data;
+
+import duke.data.task.*;
```
I think that explicitly listing the imported classes will be clearer

yzhilong commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,135 @@
+     /**
+      *  Prints no tasks found message.
+      */
+    public static void noSuchTasksFound() {
```
```suggestion
    public static void printNoSuchTasksFound() {
```
This would make your method names sound more consistent? Since you have a `printFoundTask()` method which does something very similar as this method

yzhilong commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,98 @@
+    public void runDuke() {
+        Ui.greet();
+        Scanner sc = new Scanner(System.in);
+        while(!isTerminated) {
```
```suggestion
        while (!isTerminated) {
```
I think there should be a whitespace following a java reserved word?

yzhilong commented on src/main/java/duke/ui/Ui.java:
```diff
@@ -0,0 +1,135 @@
+     * @return Error message to the user.
+     */
+    public static String dateMissing() {
+        return Messages.INDENTATION + Messages.MISSING_DATE;
```
I think that you could add another method to add indentation to a String? Since you have `Messages.INDENTATION + ...` repeated many times throughout this class.

yzhilong reviewed (commented):
I like how you abstracted common messages into a Messages class!

wonyeji committed:
Update src/main/java/duke/command/DoneCommand.java

Co-authored-by: Elizabeth Chow <51410656+lizchow@users.noreply.github.com>

wonyeji committed:
Update src/main/java/duke/data/task/Task.java

Co-authored-by: Elizabeth Chow <51410656+lizchow@users.noreply.github.com>

wonyeji committed:
Update src/main/java/duke/parser/Parser.java

Co-authored-by: Elizabeth Chow <51410656+lizchow@users.noreply.github.com>

wonyeji committed:
Resolve Checkstyle violations

wonyeji committed:
Merge branch 'master' of https://github.com/wonyeji/ip

wonyeji committed:
Add more CheckStyle changes

wonyeji committed:
Update build.gradle to support creating Jar files

wonyeji committed:
Add echo Duke GUI

wonyeji committed:
Add Images for Duke

wonyeji committed:
Add preliminary GUI

wonyeji committed:
Add JavaDoc and Fix CheckStyle errors

wonyeji committed:
Fix DeleteCommand and DoneCommand bugs and add HelpCommand

wonyeji committed:
Remove unnecessary import statements

wonyeji committed:
Add Assertions

Certain assumptions are made in the code.

Add assertions to verify these assumptions to catch bugs.

Assertion errors will be thrown if wrong assumptions were made.

wonyeji committed:
Improve Code Quality

Remove some code duplicates and deep nested methods.

Let's refactor the code to be more readable and understandable.

wonyeji committed:
Merge pull request #2 from wonyeji/branch-A-CodeQuality

Improve Code Quality

wonyeji committed:
Merge branch 'master' into branch-A-Assertions

wonyeji committed:
Merge pull request #1 from wonyeji/branch-A-Assertions

Add Assertions

wonyeji committed:
Update DoneCommand class to write to file correctly

wonyeji committed:
Add Archives function

wonyeji committed:
Improve GUI

wonyeji committed:
Enhance readability and abstraction, refactor some code

wonyeji committed:
Add product screenshot

wonyeji committed:
Update README.md

wonyeji committed:
Set theme jekyll-theme-slate

wonyeji committed:
Update README.md

wonyeji committed:
Merge branch 'master' of https://github.com/wonyeji/ip

wonyeji committed:
Update table format on README.md

wonyeji committed:
Set theme jekyll-theme-minimal

wonyeji committed:
Update README.md

wonyeji committed:
Merge branch 'master' of https://github.com/wonyeji/ip

wonyeji committed:
Update README.md

wonyeji committed:
Set theme jekyll-theme-slate

wonyeji committed:
Update README.md

wonyeji committed:
Merge branch 'master' of https://github.com/wonyeji/ip

wonyeji committed:
Update README.md
